### PR TITLE
[SPARK-43815][SQL] Add `to_varchar` alias for `to_char`

### PR DESCRIPTION
**Pull request by learningchess2003**
### What changes were proposed in this pull request?
Add the ```to_varchar``` alias for ```to_char``` to support additional SQL functions.

### Why are the changes needed?
To minimize code changes for people using Spark SQL who are used to other engines that support ```to_varchar```.

### Does this PR introduce _any_ user-facing change?
Yes. The user can now invoke ```to_varchar``` instead of ```to_char``` in their SQL applications.

### How was this patch tested?
```to_varchar``` was added in ```numeric.sql``` and a new Golden file was generated to check that the query output was correct.

### CODE REVIEW

**Comment by learningchess2003** [comment]
@dtenedor Once I am finished implementing the other functions, you can review. Of course, if there is something which you noticed I should fix earlier on, let me know!

**Comment by MaxGekk** [comment]
@learningchess2003 Do you have an account at JIRA? If not, please, submit a request:
Screenshot 2023-05-26 at 11 06 12

**Review by MaxGekk — COMMENTED** [review]
@learningchess2003 Do you have an account at JIRA? If not, please, submit a request:
Screenshot 2023-05-26 at 11 06 12

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql:887]: Please, don't use some internal ids like SC-131010

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql
Comment (learningchess2003) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql:887]: Alright, will fix that.

**Comment by MaxGekk** [comment]
@learningchess2003 Allow GitHub action in your fork, please: https://github.com/apache/spark/pull/41319/checks?check_run_id=13791306580

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql
Comment (dtenedor) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql:890]: I think you will also need to get analyzer golden file results as well. 

image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:534]: Change the version to 3.5.0. We are not going to backport this alias back to already released 3.4.0.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql:887]: Please, change the ticket title. Now it is too much generic. The PR's title looks nice for that.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Comment by learningchess2003** [comment]
Jenkins Trigger: CodeSync-RegenGoldenFiles { "test_suite": "SQL", "test_case": "numeric.sql"}

**Comment by learningchess2003** [comment]
@dtenedor @MaxGekk Addressed the comments. Let me know what you think!

**Review by dtenedor — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/numeric.sql:890]: Could you move this test out from the `postgreSQL` folder, for instance to `sql-tests/inputs/charvarchar.sql`.

The tests like this were borrowed from PostgreSQL when we tried to achieve feature parity to the DBMS:
https://issues.apache.org/jira/browse/SPARK-27763
It is better to don't add new checks there.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Comment by learningchess2003** [comment]
@MaxGekk Comments addressed.

File: sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql
Comment (MaxGekk) [inline sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:142]: Let's replace this by simpler check:
```sql
select to_varchar(date'26-May-2023', 'yyyy.mm.dd');
```

Comment (learningchess2003) [reply sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:142]: @MaxGekk  I don't think to_char/to_varchar accepts a timestamp literal yet. At the very least, when running this command, I got the following error:
```
-- !query
select to_varchar(date'26-May-2023', 'yyyy.mm.dd')
-- !query schema
struct<>
-- !query output
org.apache.spark.sql.catalyst.parser.ParseException
{
  "errorClass" : "INVALID_TYPED_LITERAL",
  "sqlState" : "42604",
  "messageParameters" : {
    "value" : "'26-May-2023'",
    "valueType" : "\"DATE\""
  },
  "queryContext" : [ {
    "objectType" : "",
    "objectName" : "",
    "startIndex" : 19,
    "stopIndex" : 35,
    "fragment" : "date'26-May-2023'"
  } ]
}
```

Comment (MaxGekk) [reply sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:142]: Yep, let's test what it supports.

Comment (learningchess2003) [reply sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:142]: Should I just stick with the original then and make it shorter?

**Comment by learningchess2003** [comment]
@MaxGekk Alright, shortened it and just had it call some standalone examples.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

File: sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql
Comment (dongjoon-hyun) [inline sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:123]: These three lines look repeating. Could you simply the above three lines into a single liner? Maybe, like the following?
```
-- to_varchar is an alias for to_char
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@learningchess2003 Could re-generate golden files:
```
$ SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/testOnly *ExpressionsSchemaSuite"
```

File: sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql
Comment (learningchess2003) [reply sql/core/src/test/resources/sql-tests/inputs/charvarchar.sql:123]: Alright, will do so.

**Comment by learningchess2003** [comment]
@MaxGekk @dongjoon-hyun Changed accordingly!

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @learningchess2003 .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (if CI passes)

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @learningchess2003 and @dongjoon-hyun @dtenedor for review.

**Comment by MaxGekk** [comment]
@learningchess2003 Congratulations with your first contribution to Apache Spark!

**Pull request closed without merging**
